src/base.jl

```diff
@@ -557,6 +557,72 @@ function get_forecasts(
     return forecasts
 end
 
+"""
+    get_forecasts(
+                  sys::System,
+                  initial_time::Dates.DateTime,
+                  components_iterator,
+                  label::Union{String, Nothing}=nothing,
+                 )
+
+# Arguments
+- `sys::System`: system
+- `initial_time::Dates.DateTime`: time designator for the forecast
+- `components_iter`: iterable (array, iterator, etc.) of Component values
+- `label::Union{String, Nothing}`: forecast label or nothing
+
+Return forecasts of any type <: Forecast that match the components and label.
+
+This method is slower than the first version because it has to compare components and label
+as well as build a new vector.
+
+Throws InvalidParameter if eltype(components_iterator) is a concrete type and no forecast is
+found for a component.
+"""
+function get_forecasts(sys::System,
+                       initial_time::Dates.DateTime,
+                       components_iterator,
+                       label::Union{String, Nothing}=nothing,
+                      )
+
+    return get_forecasts(Forecast,
+                        sys,
+                        initial_time,
+                        components_iterator,
+                        label)
+end
+
+"""
+    get_component_forecasts(::Type{T}, sys::System, initial_time::Dates.DateTime)
+
+Return an generator of forecasts for the component <: T can be concrete or abstract.
+
+Call collect on the result if an array is desired.
+
+This method is fast and efficient because it returns an iterator to existing vectors.
+
+# Examples
+```julia
+iter = PowerSystems.get_forecasts(RenewableFix, sys, initial_time)
+forecasts = PowerSystems.get_forecasts(RenewableFix, sys, initial_time) |> collect
+forecasts = collect(PowerSystems.get_forecasts(RenewableFix, sys))
+```
+
+See also: [`iterate_forecasts`](@ref)
+"""
+function get_component_forecasts(
```

Comment:
Here is a one-liner that could replace get_component_forecasts and get_components_forecasts:

```suggestion
function get_component_forecasts(
                                  ::Type{T},
                                 sys::System,
                                 initial_time::Dates.DateTime,
                                ) where T <: Component
    return (f for k in keys(sys.forecasts.data) if k.initial_time == initial_time
              for f in sys.forecasts.data[k] if get_component(f) isa T)
end
```

Reply:
There is a problem with this implementation. The return type is too complicated to do dispatch on the forecast type afterwards.

Reply:
I decided to do a collect call in PowerSimulations, this only happens once so it will be ok. If it becomes a bottle neck ir can be tackled later.